backend/cfg/cfg_prologue.ml

```diff
@@ -103,41 +103,160 @@ module Instruction_requirements = struct
+         match Instruction_requirements.basic instr with
+         | Requirements Requires_prologue -> true
+         | Prologue | Epilogue
+         | Requirements (No_requirements | Requires_no_prologue) ->
+           false)
+  ||
+  match Instruction_requirements.terminator block.terminator fun_name with
+  | Requires_prologue -> true
+  | No_requirements | Requires_no_prologue -> false
+
+let descendants (cfg : Cfg.t) (block : Cfg.basic_block) : Label.Set.t =
+  let visited = ref Label.Set.empty in
+  let rec collect label =
+    if not (Label.Set.mem label !visited)
+    then (
+      visited := Label.Set.add label !visited;
+      let block = Cfg.get_block_exn cfg label in
+      Label.Set.iter
+        (fun succ_label -> collect succ_label)
+        (Cfg.successor_labels ~normal:true ~exn:true block))
+  in
+  collect block.start;
+  !visited
+
+let can_place_prologue (prologue_label : Label.t) (cfg : Cfg.t)
+    (doms : Cfg_dominators.t) (loop_infos : Cfg_loop_infos.t)
+    (epilogue_blocks : Label.Set.t) =
+  let prologue_block = Cfg.get_block_exn cfg prologue_label in
+  (* Moving a prologue to a loop might cause it to execute multiple times, which
+     is both inefficient as well as possibly incorrect.
+
+     Having a non-zero stack offset means that the prologue is added after a
+     [Pushtrap] or [Stackoffset] which shouldn't be allowed. *)
+  if Cfg_loop_infos.is_in_loop loop_infos prologue_label
+     || prologue_block.stack_offset <> 0
+  then false
+  else
+    (* Check that the blocks requiring an epilogue are dominated by the prologue
+       block. Consider the following CFG:
+
+     *  Block A: Condition with branch to Block B / C
+     *  Block B: Contains an instruction requiring a prologue, with 
+        terminator that jumps to Block C
+     *  Block C: Return
+
+       We have the choice of putting the prologue in Block A or B, and we would
+       place an epilogue in Block C.
+
+       If we try to place the prologue in block B, the prologue would not
+       dominate the epilogue in block C, so in some cases the epilogue would be
+       executed without a prologue on the stack, which would be illegal. *)
+    Label.Set.for_all
+      (fun epilogue_label ->
+        Cfg_dominators.is_dominating doms prologue_label epilogue_label)
+      epilogue_blocks
+
+let find_prologue_and_epilogues_shrink_wrapped (cfg : Cfg.t) =
+  let rec visit (tree : Cfg_dominators.dominator_tree) (cfg : Cfg.t)
+      (doms : Cfg_dominators.t) (loop_infos : Cfg_loop_infos.t) =
+    let block = Cfg.get_block_exn cfg tree.label in
+    let epilogue_blocks =
+      Label.Set.filter
+        (fun label ->
+          let block = Cfg.get_block_exn cfg label in
+          let fun_name = Cfg.fun_name cfg in
+          match
+            Instruction_requirements.terminator block.terminator fun_name
+          with
+          | Requires_no_prologue -> true
+          | No_requirements | Requires_prologue -> false)
+        (descendants cfg block)
```

Comment:
This recursive call can get expensive, it allocates a lot and recomputes the epilogue_blocks every time.